feat(core): add opt-in test module teardown configuration

crisbeto opened the pull request:
We currently have two long-standing issues related to how `TestBed` tests are torn down:
1. The dynamically-created test module isn't going to be destroyed, preventing the `ngOnDestroy` hooks on providers from running and keeping the component `style` nodes in the DOM.
2. The test root elements aren't going to be removed from the DOM. Instead, they will be removed whenever another test component is created.

By themselves, these issues are easy to resolve, but given how long they've been around, there are a lot of unit tests out there that depend on the broken behavior.

These changes address the issues by introducing APIs that allow users to opt into the correct test teardown behavior either at the application level via `TestBed.initTestEnvironment` or the test suite level via `TestBed.configureTestingModule`.

At the moment, the new teardown behavior is opt-in, but the idea is that we'll eventually make it opt-out before removing the configuration altogether.

Fixes #18831.

crisbeto commented on packages/router/src/router_module.ts:
```diff
@@ -7,7 +7,7 @@
 
 import {APP_BASE_HREF, HashLocationStrategy, Location, LOCATION_INITIALIZED, LocationStrategy, PathLocationStrategy, PlatformLocation, ViewportScroller} from '@angular/common';
-import {ANALYZE_FOR_ENTRY_COMPONENTS, APP_BOOTSTRAP_LISTENER, APP_INITIALIZER, ApplicationRef, Compiler, ComponentRef, Inject, Injectable, InjectionToken, Injector, ModuleWithProviders, NgModule, NgModuleFactoryLoader, NgProbeToken, Optional, Provider, SkipSelf, SystemJsNgModuleLoader} from '@angular/core';
+import {ANALYZE_FOR_ENTRY_COMPONENTS, APP_BOOTSTRAP_LISTENER, APP_INITIALIZER, ApplicationRef, Compiler, ComponentRef, Inject, Injectable, InjectionToken, Injector, ModuleWithProviders, NgModule, NgModuleFactoryLoader, NgProbeToken, OnDestroy, Optional, Provider, SkipSelf, SystemJsNgModuleLoader} from '@angular/core';
```
The changes in this file are also included in https://github.com/angular/angular/pull/42560. I copied them in here for now, because the error that was being thrown was adding unnecessary noise to the logs. Once https://github.com/angular/angular/pull/42560 goes in, I'll remove these changes.

crisbeto commented on packages/core/test/test_bed_spec.ts:
```diff
@@ -1284,3 +1285,185 @@ describe('TestBed', () => {
+    expect(fixtureDocument.body.contains(fixture.nativeElement)).toBe(false);
+  });
+
+  it('should re-throw errors that were thrown during fixture cleanup', () => {
```
One note here: the original design only called for test **module** errors to be re-thrown, **not test component**, but I decided to make this change, because I was looking at how the Components repo is working around this and I realized that not throwing errors for errors during test component destruction can still lead to difficult to trace down bugs. For reference: https://github.com/angular/components/blob/master/test/karma-test-shim.js#L61

crisbeto committed:
feat(core): add opt-in test module teardown configuration

We currently have two long-standing issues related to how `TestBed` tests are torn down:
1. The dynamically-created test module isn't going to be destroyed, preventing the `ngOnDestroy` hooks on providers from running and keeping the component `style` nodes in the DOM.
2. The test root elements aren't going to be removed from the DOM. Instead, they will be removed whenever another test component is created.

By themselves, these issues are easy to resolve, but given how long they've been around, there are a lot of unit tests out there that depend on the broken behavior.

These changes address the issues by introducing APIs that allow users to opt into the correct test teardown behavior either at the application level via `TestBed.initTestEnvironment` or the test suite level via `TestBed.configureTestingModule`.

At the moment, the new teardown behavior is opt-in, but the idea is that we'll eventually make it opt-out before removing the configuration altogether.

Fixes #18831.

AndrewKushnir commented on packages/core/testing/src/test_hooks.ts:
```diff
@@ -13,18 +13,29 @@
+  _global.beforeEach(getCleanupHook(false));
+}
+
+if (_global.afterEach) {
+  _global.afterEach(getCleanupHook(true));
+}
```
Could you plz add a quick note here on why we need to provide `afterEach` callback too?

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -402,17 +433,67 @@ export class TestBedRender3 implements TestBed {
+        });
+      }
+    } finally {
+      testRenderer.removeAllRootElements();
```
Given that the `TestComponentRenderer` class can be configured via DI, it may not have the `removeAllRootElements` function:
```suggestion
      testRenderer.removeAllRootElements?.();
```

crisbeto replied:
I'll mark it as optional, although this interface is private.

AndrewKushnir commented on packages/core/testing/src/test_bed_common.ts:
```diff
@@ -41,8 +42,29 @@ export type TestModuleMetadata = {
+ * @publicApi
+ */
+export interface ModuleTeardownOptions {
+  /** Whether the test module should be destroyed after every test. */
```
It'd be great to mention a default value here too.

crisbeto replied:
I avoid mentioning default values in docs, because they inevitably go out of sync when the default changes.

AndrewKushnir replied:
Ok, got it. I just noticed that you've added the default value description to the `rethrowErrors` field, but not this one.

crisbeto replied:
I don't anticipate for the `rethrowErrors` default value to change, whereas the teardown one will.

AndrewKushnir commented on packages/core/testing/src/test_bed_common.ts:
```diff
@@ -41,8 +42,29 @@ export type TestModuleMetadata = {
+}
+
+/**
+ * Object used to configure the test module teardown behavior in `TestBed`.
```
I think we'd need to add more info here to explain why these options exist and provide recommendations on when they should be used/changed and what the plan is (that these options will eventually be deprecated and removed, so we do not encourage using them unless it's absolutely needed and blocking tests).

crisbeto replied:
The idea was to have a silent rollout of this in 12.1.0 and then add a page in the docs similar to https://angular.io/guide/migration-update-libraries-tslib.

AndrewKushnir replied:
Thanks for additional info, I'd propose adding a couple sentences to describe that:
```suggestion
 * Object used to configure the test module teardown behavior in `TestBed`.
 * This option is exposed as a public API to allow switching to the legacy TestBed behavior temporarily
 * (where testing module was not removed as a part of the test teardown process) in case your tests rely on
 * this old behavior. These config options will be deprecated and removed in the future versions of Angular.
```

crisbeto replied:
I'm not sure that people who aren't familiar with the `TestBed` internals would know what this is referring to.

AndrewKushnir replied:
We can update the message to make it more user-friendly. I'd like to include some information about the fact that these config options are temporary and developers should not rely on them.

crisbeto replied:
IMO we should capture the fact that it's temporary when we eventually add the page on AIO. Also the plan is to provide schematics that add the parameter and then remove it once we've changed the default value so hopefully people won't have to deal with it themselves.

AndrewKushnir replied:
The `ModuleTeardownOptions` would become discoverable on the AIO once this PR is merged, see
https://pr42566-893884a.ngbuilds.io/api/core/testing/ModuleTeardownOptions.

crisbeto replied:
I meant that providing all the context behind the option would take a few paragraphs of text which seems like overkill for an API doc. E.g. the design doc for these changes takes 170 words to explain the issues.

AndrewKushnir reviewed (commented):
Looks great, thanks @crisbeto! I've just left a few comments, please have a look when you get a chance.

josephperrott reviewed (approved):
LGTM

josephperrott commented on packages/core/testing/src/test_hooks.ts:
```diff
@@ -13,18 +13,29 @@
+
+function getCleanupHook(expectedTeardownValue: boolean) {
+  return () => {
+    if ((TestBed as unknown as TestBedInternal).shouldTearDownTestingModule() ===
```
Because `resetTestingModule` now relies on `TestBed` already being initialized before we can call it, we end up seeing alot of: 

```
Error: Need to call TestBed.initTestEnvironment() first
```

crisbeto commented:
@AndrewKushnir I've addressed the feedback.

@josephperrott I've pushed a fix for it. I had to revert one place where I tried to make something a couple lines shorter, but it ended up causing the error.

crisbeto committed:
fixup! feat(core): add opt-in test module teardown configuration

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -402,17 +448,68 @@ export class TestBedRender3 implements TestBed {
+    }
+
+    // Otherwise use the configured behavior or default to rethrowing.
+    return instanceOptions?.rethrowErrors ?? environmentOptions?.rethrowErrors ?? true;
```
Would it make sense to extract this default value to a const too (similar to `TEARDOWN_TESTING_MODULE_ON_DESTROY_DEFAULT`)?

crisbeto replied:
I think the constant made sense for the teardown option since we're going to change it over time. I don't anticipate for the error rethrowing default behavior to change.

AndrewKushnir commented on packages/core/testing/src/r3_test_bed.ts:
```diff
@@ -402,17 +448,68 @@ export class TestBedRender3 implements TestBed {
+        });
+      }
+    } finally {
+      testRenderer.removeAllRootElements?.();
```
Thanks for adding the safe function call. I've checked g3 codebase and there is a couple places where a custom implementation of the `TestComponentRenderer` is provided (thus the `removeAllRootElements` would not be present).

atscott reviewed (approved):
reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

mary-poppins commented:
You can preview 0a272fd at https://pr42566-0a272fd.ngbuilds.io/.
You can preview dd2cb6e at https://pr42566-dd2cb6e.ngbuilds.io/.
You can preview b7a29d2 at https://pr42566-b7a29d2.ngbuilds.io/.
You can preview 893884a at https://pr42566-893884a.ngbuilds.io/.

mary-poppins commented:
You can preview 0a272fd at https://pr42566-0a272fd.ngbuilds.io/.
You can preview dd2cb6e at https://pr42566-dd2cb6e.ngbuilds.io/.
You can preview b7a29d2 at https://pr42566-b7a29d2.ngbuilds.io/.
You can preview 893884a at https://pr42566-893884a.ngbuilds.io/.

jessicajaniuk reviewed (approved):
LGTM 🍪 

Reviewed-for: public-api

crisbeto committed:
fixup! feat(core): add opt-in test module teardown configuration
